apps/core/secureaction.cpp

```diff
@@ -0,0 +1,254 @@
+                qWarning() << "Error querying" << deviceLockService << "on the system bus";
+            } else if (reply.value()) {
+                m_available = true;
+                emit availableChanged(true);
+            } else {
+                qWarning() << deviceLockService << "is not available on the system bus";
+            }
+        });
+    }
+
+    static std::shared_ptr<ServiceWatcher> instance()
+    {
+        static std::weak_ptr<ServiceWatcher> instance;
+
+        std::shared_ptr<ServiceWatcher> watcher = instance.lock();
+
+        if (!watcher) {
+            watcher = std::make_shared<ServiceWatcher>();
+
+            instance = watcher;
+        }
+
+        return watcher;
+    }
+
+    bool isAvailable() const { return m_available; }
+
+signals:
+    void availableChanged(bool available);
+
+private:
+    bool m_available = false;;
+};
+
+SecureActionAuthenticatorAdaptor::SecureActionAuthenticatorAdaptor(SecureAction *action)
+    : QDBusAbstractAdaptor(action)
+    , m_secureAction(action)
+{
+}
+
+void SecureActionAuthenticatorAdaptor::PermissionGranted(uint)
+{
+    if (m_secureAction->m_authenticating) {
+        QJSValue resolve = m_secureAction->m_resolve;
+
+        m_secureAction->m_authorized = true;
+        m_secureAction->m_authenticating = false;
+        m_secureAction->m_resolve = QJSValue();
+
+        resolve.call();
+    }
+}
+
+void SecureActionAuthenticatorAdaptor::Aborted()
+{
+    m_secureAction->m_authenticating = false;
+    m_secureAction->m_resolve = QJSValue();
+}
+
+SecureAction::SecureAction(QObject *parent)
+    : QObject(parent)
+    , m_serviceWatcher(ServiceWatcher::instance())
+    , m_replyPath(replyPath())
+    , m_available(m_serviceWatcher->isAvailable())
+{
+    connect(m_serviceWatcher.get(), &ServiceWatcher::availableChanged, this, &SecureAction::setAvailable);
+}
+
+SecureAction::~SecureAction()
+{
+    cancel();
+}
+
+bool SecureAction::isAvailable() const
+{
+    return m_available;
+}
+
+QString SecureAction::message() const
+{
+    return m_message;
+}
+
+void SecureAction::setMessage(const QString &message)
+{
+    if (m_message != message) {
+        m_message = message;
+
+        m_authorized = false;
+        cancel();
+
+        emit messageChanged();
+    }
+}
+
+void SecureAction::perform(const QJSValue &resolve)
+{
+    if (m_authorized) {
+        QJSValue(resolve).call();
+    } else if (m_available && !m_message.isEmpty()) {
+        m_resolve = resolve;
+
+        if (m_authenticating) {
+            return;
+        }
+
+        QDBusConnection systemBus = QDBusConnection::systemBus();
+
+        if (!m_registered) {
+            m_registered = systemBus.registerObject(m_replyPath, this);
+
+            if (!m_registered) {
+                qmlInfo(this) << "DBus adaptor registration error " << systemBus.lastError().name() << " " << systemBus.lastError().message();
+                return;
+            }
+        }
+
+        m_authenticating = true;
+
```

Comment:
What ensures that SecureActionAuthenticatorAdaptor::PermissionGranted() will be invoked by the deviceLockService as a result of this call?  Is there some connection to a dbus signal which I'm missing?

Reply:
Or is it just documented that the deviceLockService will invoke that method on the replyPath?  If so, maybe a comment could be added above to state this.

Reply:
Yes, the service will send a reply to the reply path object.

Reply:
Added a comment.